api/workload/v2/stream.go

```diff
@@ -0,0 +1,168 @@
+	var err error
+	done := make(chan struct{})
+	go func() {
+		res, err = s.client.Recv()
+		close(done)
+	}()
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-done:
+		return res, err
+	}
+}
+
+// Close closes the stream and the underlying connection.
+func (s *managedStream) Close() error {
+	var errs []string
+	if err := s.client.CloseSend(); err != nil {
+		errs = append(errs, err.Error())
+	}
+	if err := s.closer.Close(); err != nil {
+		errs = append(errs, err.Error())
+	}
+	if errs != nil {
+		return errors.New(strings.Join(errs, "; "))
+	}
+	return nil
+}
+
+func newStreamManager(ctx context.Context, logger *logrus.Logger, addr string) (*streamManager, error) {
+	if !strings.HasPrefix(addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", addr)
+	}
+	return &streamManager{
+		StreamChan:     make(chan *managedStream, 1),
+		ctx:            ctx,
+		logger:         logger,
+		addr:           addr,
+		reconnectChan:  make(chan struct{}, 1),
+		ConnectionChan: make(chan bool, 1),
+	}, nil
+}
+
+// Reconect informs the stream manager that the current stream is unusable.
+func (c *streamManager) Reconnect() {
+	c.ConnectionChan <- false
+	c.reconnectChan <- struct{}{}
+}
+
+// Stop stops the stream manager
+func (c *streamManager) Stop() {
+	c.logger.Debug("Shutting down stream manager.")
+	if c.StreamChan != nil {
+		close(c.StreamChan)
+	}
+	if c.ConnectionChan != nil {
+		close(c.ConnectionChan)
+		c.logger.WithField("queued", len(c.ConnectionChan)).Debug("Emptying connection chan.")
+		for range c.ConnectionChan {
+		}
+		c.ConnectionChan = nil
+	}
+}
+
+// Start starts the stream manager in the background.
+func (c *streamManager) Start() {
+	c.reconnectChan <- struct{}{}
+	go c.start()
+}
+
+func (c *streamManager) start() {
+	for {
+		select {
+		case _, ok := <-c.reconnectChan:
+			if ok {
+				stream, err := c.newStream(c.ctx, c.addr)
+				if err != nil {
+					continue
+				}
+				c.StreamChan <- stream
+				c.ConnectionChan <- true
+				c.logger.Debug("Created updated stream")
+			}
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *streamManager) newStream(ctx context.Context, addr string) (*managedStream, error) {
+	backoff := newBackoff()
+	for {
+		var stream workload.SpiffeWorkloadAPI_FetchX509SVIDClient
+		var err error
+		conn, err := newConn(ctx, addr)
+		if err != nil {
+			goto retry
+		}
+		stream, err = newX509SVIDStream(ctx, conn)
+		if err == nil {
+			return &managedStream{
+				client: stream,
+				closer: conn,
+			}, nil
+		}
+	retry:
+		c.logger.WithError(err).Debug("Error creating stream, retrying.")
+		select {
+		case <-ctx.Done():
+			c.logger.Debug("Stream creator shutting down.")
+			return nil, ctx.Err()
+		case <-time.After(backoff.Duration()):
+		}
+	}
+}
+
+func newConn(ctx context.Context, addr string) (*grpc.ClientConn, error) {
+	conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure())
```

Comment:
this should also have `grpc.WithBlock()` since its managing all the connection details. Having grpc dialing in the background is confusing.